plumbing/format/packfile/scanner.go

```diff
@@ -378,110 +378,82 @@ func (s *Scanner) Close() error {
+//   isn't an io.SeekReader, but we still want to know the current offset.
+// - Writes to the hash writer what it reads, with the aid of a smaller buffer.
+//   The buffer helps avoid a performance penality for performing small writes
+//   to the crc32 hash writer.
+type scannerReader struct {
+	reader io.Reader
+	crc    io.Writer
+	rbuf   *bufio.Reader
+	wbuf   *bufio.Writer
+	offset int64
 }
 
-// Read reads up to len(p) bytes into p.
-func (r *trackableReader) Read(p []byte) (n int, err error) {
-	n, err = r.Reader.Read(p)
-	r.count += int64(n)
-
-	return
-}
-
-// Seek only supports io.SeekCurrent, any other operation fails
-func (r *trackableReader) Seek(offset int64, whence int) (int64, error) {
-	if whence != io.SeekCurrent {
-		return -1, ErrSeekNotSupported
+func newScannerReader(r io.Reader, h io.Writer) *scannerReader {
+	sr := &scannerReader{
+		reader: r,
+		crc:    h,
 	}
+	sr.rbuf = bufio.NewReader(sr.reader)
+	sr.wbuf = bufio.NewWriterSize(sr.crc, 64)
 
-	return r.count, nil
+	return sr
 }
 
-func newByteReadSeeker(r io.ReadSeeker) *bufferedSeeker {
-	return &bufferedSeeker{
-		r:      r,
-		Reader: *bufio.NewReader(r),
-	}
+func (r *scannerReader) Reset(reader io.Reader) {
+	r.reader = reader
+	r.offset = 0
+	r.rbuf.Reset(r.reader)
+	r.wbuf.Reset(r.crc)
 }
 
-type bufferedSeeker struct {
-	r io.ReadSeeker
-	bufio.Reader
-}
+func (r *scannerReader) Read(p []byte) (n int, err error) {
+	n, err = r.rbuf.Read(p)
 
-func (r *bufferedSeeker) Seek(offset int64, whence int) (int64, error) {
-	if whence == io.SeekCurrent && offset == 0 {
-		current, err := r.r.Seek(offset, whence)
-		if err != nil {
-			return current, err
-		}
-
-		return current - int64(r.Buffered()), nil
-	}
-
-	defer r.Reader.Reset(r.r)
-	return r.r.Seek(offset, whence)
-}
-
-type readerResetter interface {
-	io.ReadCloser
-	zlib.Resetter
+	r.offset += int64(n)
+	return r.wbuf.Write(p[:n])
 }
 
-type reader interface {
-	io.Reader
-	io.ByteReader
-	io.Seeker
-}
-
-type teeReader struct {
-	reader
-	w         hash.Hash32
-	bufWriter *bufio.Writer
+func (r *scannerReader) ReadByte() (b byte, err error) {
+	b, err = r.rbuf.ReadByte()
+	if err == nil {
+		r.offset++
+		return b, r.wbuf.WriteByte(b)
+	}
+	return
 }
 
-func newTeeReader(r reader, h hash.Hash32) *teeReader {
-	return &teeReader{
-		reader:    r,
-		w:         h,
-		bufWriter: bufio.NewWriter(h),
-	}
+func (r *scannerReader) Flush() error {
+	return r.wbuf.Flush()
 }
 
-func (r *teeReader) Read(p []byte) (n int, err error) {
-	r.Flush()
+// Seek seeks to a location. If the underlying reader is not an io.ReadSeeker,
+// then only whence=io.SeekCurrent is supported, any other operation fails.
+func (r *scannerReader) Seek(offset int64, whence int) (int64, error) {
+	var err error
 
-	n, err = r.reader.Read(p)
-	if n > 0 {
-		if n, err := r.w.Write(p[:n]); err != nil {
-			return n, err
+	if seeker, ok := r.reader.(io.ReadSeeker); !ok {
+		if whence != io.SeekCurrent && offset != 0 {
```

Comment:
should be `||`

Reply:
Thanks! I've just noticed the last second modification I made that's breaking the tests too. Will push shortly :)

Reply:
Got there in the end :)